lib/charms/opensearch/v0/helper_conf_setter.py

```diff
@@ -240,33 +240,37 @@ def __deep_delete(self, source, node_keys: List[str]):
         if source is None:
             return
 
-        leaf_level = self.__leaf_level(source, node_keys)
-        leaf_key = node_keys.pop(0)
-
-        if leaf_key in leaf_level and (
-            isinstance(leaf_level[leaf_key], Mapping) or not leaf_key.startswith("[")
-        ):
-            del leaf_level[leaf_key]
-            return
-
-        if leaf_key.startswith("{"):
-            leaf_level.remove(leaf_key[1:-1])
-            return
-
-        # list
-        target_index = self.__target_array_index(leaf_level, leaf_key)
-        del leaf_level[target_index]
-
-    def __leaf_level(self, current, node_names: List[str]):
+        try:
+            leaf_container = self.__leaf_container(source, node_keys)
+            leaf_key = node_keys.pop(0)
+
+            # remove simple type elements and entire collections by key
+            if leaf_key in leaf_container and leaf_key[0] not in {"{", "["}:
+                del leaf_container[leaf_key]
+                return
+
+            # remove element from set
+            if leaf_key.startswith("{"):
+                leaf_container.remove(leaf_key[1:-1])
+                return
+
+            # remove element from list
+            target_index = self.__target_array_index(leaf_container, leaf_key)
+            del leaf_container[target_index]
+        except AttributeError:
+            # element not found
+            pass
+
+    def __leaf_container(self, current, node_names: List[str]):
```

Comment:
I can't remember if this was mentioned before, if so close this.

Python convention for private methods is single `_`, as dunder `__` is reserved for builtins. Probably too much of a faff to change now, but worth mentioning if there isn't an explicit reason to have a double-under here.

Reply:
Yeah so I got into the habit of adding `__` as a way to say "class private" and no way to access it (as python name mangle methods prefixed with a double underscore), you can find more details [here](https://stackoverflow.com/questions/1301346/what-is-the-meaning-of-single-and-double-underscore-before-an-object-name).
However, I agree that it probably is overkill and not needed in this context - will remove those in a next PR